addon/utils/disposable.js

```diff
@@ -0,0 +1,89 @@
+import Ember from 'ember';
+import { assert } from '@ember/debug';
+import { WILL_DESTROY_PATCHED } from '../utils/flags';
+
+const { WeakMap } = Ember;
+
+/**
+ * A map of instances/array of disposables. Only exported for
+ * testing purposes.
+ *
+ * @public
+ */
+export let registeredDisposables = new WeakMap();
+
+/**
+ * Registers a new disposable function to run for an instance. Will
+ * handle lazily creating a new array to store the disposables per
+ * instance if one does not exist. Will additionally patch an object's
+ * `willDestroy` hook to ensure the destroyables are run/destroyed when
+ * the object is destroyed.
+ *
+ * @public
+ * @param {*} obj the instance to store the disposable for
+ * @param {*} dispose a function that disposes of instance resources
+ * @returns a disposable object, which wraps the passed in `dispose` function
+ */
+export function registerDisposable(obj, dispose) {
+  assert(
+    'Called `registerDisposable` where `obj` is not an object',
+    typeof obj === 'object'
+  );
+  assert(
+    'Called `registerDisposable` where `dispose` is not a function',
+    typeof dispose === 'function'
+  );
+  assert(
+    'Called `registerDisposable` without implementing `willDestroy` that calls `runDisposables`',
+    !!obj[WILL_DESTROY_PATCHED]
+  );
+
+  let disposables = getRegisteredDisposables(obj);
+  let disposable = _toDisposable(dispose);
+
+  disposables.push(disposable);
+
+  return disposable;
+}
+
+/**
+ * Runs all disposables for a given instance.
+ *
+ * @public
+ * @param {*} obj the instance to run the disposables for
+ */
+export function runDisposables(obj) {
+  let disposables = registeredDisposables.get(obj);
+
+  if (!disposables) {
+    return;
+  }
+
+  for (let i = 0, l = disposables.length; i < l; i++) {
+    let disposable = disposables.pop();
```

Comment:
we should likely not mutate the array while we are iterating it. That will force extra bounds checks for no reason.
We should not also not both caching length.

The following  would be both performant and safe.

```js
let disposables  = registeredDisposables.get(obj);
registeredDisposables.delete(obj); // or set to `null`

for (let i =0; i<disposables.length; i++) {
  disposables[i].dispose(); // or simplified, without a disposable wrapper object `disposables[i]();`
}
```

Reply:
Length was cached due to its mutation during the loop. I could also have used a decremented while loop, which would have iterated from end to start (and I believe would be more performant).

> That will force extra bounds checks for no reason.

Not sure what this statement means. 

I can definitely simplify this by not popping.